### PR TITLE
fix: Add cacheing CMM examples

### LINKED ISSUE #23: Need data key caching example
Before release, or as a fast-follow, we need an example in JavaScript that is analogous to the ones in Java and Python. 

https://docs.aws.amazon.com/encryption-sdk/latest/developer-guide/sample-cache-example.html

### PR DESCRIPTION
**Pull request by seebees**
resolves #23

Add examples for both node and browser.
Refactor the browser examples to be tested with karma.
Update the browser html so that it still works.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

# Check any applicable:
- [ ] Were any files moved? Moving files changes their URL, which breaks all hyperlinks to the files.

### CODE REVIEW

File: modules/example-browser/src/caching_materials_manager.ts [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+  })
+
+  /* Encryption context is a *very* powerful tool for controlling and managing access.
+   * It is ***not*** secret!
+   * Encrypted data is opaque.
+   * You can use an encryption context to assert things about the encrypted data.
+   * Just because you can decrypt something does not mean it is what you expect.
+   * For example, if you are are only expecting data from 'us-west-2',
+   * the origin can identify a malicious actor.
+   * See: https://docs.aws.amazon.com/encryption-sdk/latest/developer-guide/concepts.html#encryption-context
```
Comment (juneb) [inline modules/example-browser/src/caching_materials_manager.ts]: Better:  The following doc describes the use of the encryption context in data key caching.
https://docs.aws.amazon.com/encryption-sdk/latest/developer-guide/data-caching-details.html#caching-encryption-context

File: modules/example-browser/src/caching_materials_manager.ts [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+} from '@aws-crypto/client-browser'
+import { toBase64 } from '@aws-sdk/util-base64-browser'
+
+/* This is injected by webpack.
+ * The webpack.DefinePlugin will replace the values when bundling.
+ * The credential values are pulled from @aws-sdk/credential-provider-node
+ * Use any method you like to get credentials into the browser.
+ * See kms.webpack.config
+ */
+declare const credentials: {accessKeyId: string, secretAccessKey:string, sessionToken:string }
```
Comment (juneb) [inline modules/example-browser/src/caching_materials_manager.ts]: Do you need to add the sessionToken to all browser examples, or just this one?

Comment (seebees) [reply modules/example-browser/src/caching_materials_manager.ts]: All of them.

File: modules/example-browser/src/caching_materials_manager.ts (hunk 1 above)
Comment (seebees) [reply modules/example-browser/src/caching_materials_manager.ts]: That link is not about cacheing, it is about encryption context.  i.e. why you should include/use it.

File: modules/example-browser/src/caching_materials_manager.ts [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+  getClient,
+  encrypt,
+  decrypt,
+  WebCryptoCachingMaterialsManager,
+  getLocalCryptographicMaterialsCache
+} from '@aws-crypto/client-browser'
+import { toBase64 } from '@aws-sdk/util-base64-browser'
+
+/* This is injected by webpack.
+ * The webpack.DefinePlugin will replace the values when bundling.
```
Comment (seebees) [inline modules/example-browser/src/caching_materials_manager.ts]: Update comment, this is no longer accurate.

File: modules/example-browser/src/caching_materials_manager.ts [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+  /* Create a cache to hold the material.
+   * In this case we use the local cache provided by the Encryption SDK.
+   * The number is the maximum number of entries that will be cached.
+   * Both encrypt and decrypt requests count independently towards this total.
+   * Elements will be actively removed from the cache.
+   * The default frequency is to check one item every minute.
+   * This can be configure by passing a `proactiveFrequency`
+   * as the second paramter to however often you want to check in milliseconds.
+   */
+  const cache = getLocalCryptographicMaterialsCache(100)
```
Comment (juneb) [inline modules/example-browser/src/caching_materials_manager.ts]: A capacity value of 100 is very high. The typical value is 10. Also, I would make this a variable so it's meaning is clear. 

    const maxSize = 10
    const cache = getLocalCryptographicMaterialsCache(maxSize)

File: modules/example-browser/src/caching_materials_manager.ts [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+   */
+  const maxBytesEncrypted = 100
+
+  /* The maximum number of messages that will be encrypted under a single data key.
+   * This value is optional,
+   * but you should configure the lowest value possible.
+   */
+  const maxMessagesEncrypted = 100
+
+  const cmm = new WebCryptoCachingMaterialsManager({
```
[the remark this answers is not in the document]
Comment (seebees) [reply modules/example-browser/src/caching_materials_manager.ts]: kk

File: modules/example-browser/src/caching_materials_manager.ts [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+   * This value is optional,
+   * but you should configure the lowest value possible.
+   */
+  const maxBytesEncrypted = 100
+
+  /* The maximum number of messages that will be encrypted under a single data key.
+   * This value is optional,
+   * but you should configure the lowest value possible.
+   */
+  const maxMessagesEncrypted = 100
```
[the remark this answers is not in the document]
Comment (seebees) [reply modules/example-browser/src/caching_materials_manager.ts]: kk

File: modules/example-browser/src/caching_materials_manager.ts [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+ *     http://aws.amazon.com/apache2.0/
+ *
+ * or in the "license" file accompanying this file. This file is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* This is a simple example of using a KMS Keyring
+ * to encrypt and decrypt using the AWS Encryption SDK for Javascript in a browser.
```
[the remark this answers is not in the document]
Comment (seebees) [reply modules/example-browser/src/caching_materials_manager.ts]: yes, it is

File: modules/example-browser/src/caching_materials_manager.ts (hunk 4 above)
Comment (seebees) [reply modules/example-browser/src/caching_materials_manager.ts]: will update in both places but I will use capacity

File: modules/example-browser/src/caching_materials_manager.ts (hunk 3 above)
Comment (seebees) [reply modules/example-browser/src/caching_materials_manager.ts]: fixed

File: modules/example-browser/src/caching_materials_manager.ts [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+ *
+ *     http://aws.amazon.com/apache2.0/
+ *
+ * or in the "license" file accompanying this file. This file is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* This is a simple example of using a Caching CMM with a KMS Keyring
```
Comment (seebees) [inline modules/example-browser/src/caching_materials_manager.ts]:
```suggestion
/* This is a simple example of using a caching CMM with a KMS keyring
```

File: modules/example-browser/src/caching_materials_manager.ts [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+
+  /* The KMS keyring must be configured with the desired CMKs */
+  const keyring = new KmsKeyringBrowser({ clientProvider, generatorKeyId, keyIds })
+
+  /* Create a cache to hold the material.
+   * In this case we use the local cache provided by the Encryption SDK.
+   * The number is the maximum number of entries that will be cached.
+   * Both encrypt and decrypt requests count independently towards this total.
+   * Elements will be actively removed from the cache.
+   * The default frequency is to check one item every minute.
```
Comment (juneb) [inline modules/example-browser/src/caching_materials_manager.ts]:
```suggestion
   * By default, the SDK checks one item in the cache every 60 seconds (60,000 milliseconds).
```

File: modules/example-browser/src/caching_materials_manager.ts [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+   * This can be configure by passing a `proactiveFrequency`
```
Comment (juneb) [inline modules/example-browser/src/caching_materials_manager.ts]:
```suggestion
   * To change this frequency, pass in a `proactiveFrequency` value
```

File: modules/example-browser/src/caching_materials_manager.ts [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+   * as the second paramter to however often you want to check in milliseconds.
```
Comment (juneb) [inline modules/example-browser/src/caching_materials_manager.ts]:
```suggestion
   * as the second parameter. This value is in milliseconds.
```

File: modules/example-browser/src/caching_materials_manager.ts [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+   */
+  const capacity = 100
+  const cache = getLocalCryptographicMaterialsCache(capacity)
+
+  /* The partition name lets multiple caching CMMs share the same local cryptographic cache.
+   * If you want these CMMs to all cache the same items,
```
Comment (juneb) [inline modules/example-browser/src/caching_materials_manager.ts]:
```suggestion
   * By default, the entries for each CMM are cached separately. However, if you want these CMMs to share the cache,
```

File: modules/example-browser/src/caching_materials_manager.ts [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+   * make the partition name the same.
+   * If no partition is supplied a random one will be generated.
```
Comment (juneb) [inline modules/example-browser/src/caching_materials_manager.ts]:
```suggestion
   * If you don't supply a partition name, the Encryption SDK generates a random name.
```

File: modules/example-browser/src/caching_materials_manager.ts [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+   * This is so that sharing elements in the cache MUST be an intentional operation.
```
Comment (juneb) [inline modules/example-browser/src/caching_materials_manager.ts]:
```suggestion
   * As a result, sharing elements in the cache MUST be an intentional operation.
```

File: modules/example-browser/src/caching_materials_manager.ts [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+  const partition = 'local partition name'
+
+  /* maxAge is the time in milliseconds that an entry will be cached.
+   * Elements are actively removed from the cache.
+   */
+  const maxAge = 1000 * 60
+
+  /* The maximum amount of bytes that will be encrypted under a single data key.
+   * This value is optional,
+   * but you should configure the lowest value possible.
```
Comment (juneb) [inline modules/example-browser/src/caching_materials_manager.ts]:
```suggestion
   * but you should configure the lowest practical value.
```
The lowest possible value is 0 or minInt or whatever. That's not what we mean.

File: modules/example-browser/src/caching_materials_manager.ts [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+    maxAge,
+    maxBytesEncrypted,
+    maxMessagesEncrypted
+  })
+
+  /* Encryption context is a *very* powerful tool for controlling and managing access.
+   * It is ***not*** secret!
+   * Encrypted data is opaque.
+   * You can use an encryption context to assert things about the encrypted data.
+   * Just because you can decrypt something does not mean it is what you expect.
```
Comment (juneb) [inline modules/example-browser/src/caching_materials_manager.ts]:
```suggestion
   * The encryption context helps you to determine whether the ciphertext you retrieved is the ciphertext you expect to decrypt.
```

File: modules/example-browser/src/caching_materials_manager.ts [hunk 17]
```diff
@@ -0,0 +1,201 @@
...
+  const cachingCMM = new WebCryptoCachingMaterialsManager({
+    backingMaterials: keyring,
+    cache,
+    partition,
+    maxAge,
+    maxBytesEncrypted,
+    maxMessagesEncrypted
+  })
+
+  /* Encryption context is a *very* powerful tool for controlling and managing access.
```
Comment (juneb) [inline modules/example-browser/src/caching_materials_manager.ts]:
```suggestion
  /* Encryption context is a *very* powerful tool for controlling and managing access. When you pass an encryption context to the encrypt function, the encryption context is cryptographically bound to the data. If you don't pass in the same encryption context when decrypting, the decrypt function fails.
```

File: modules/example-browser/src/caching_materials_manager.ts [hunk 18]
```diff
@@ -0,0 +1,201 @@
...
+   * It is ***not*** secret!
```
Comment (juneb) [inline modules/example-browser/src/caching_materials_manager.ts]:
```suggestion
   * The encryption context is ***not*** secret!
```

File: modules/example-browser/src/caching_materials_manager.ts [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+   * Encrypted data is opaque.
+   * You can use an encryption context to assert things about the encrypted data.
+   * Just because you can decrypt something does not mean it is what you expect.
+   * For example, if you are are only expecting data from 'us-west-2',
+   * the origin can identify a malicious actor.
```
Comment (juneb) [inline modules/example-browser/src/caching_materials_manager.ts]:
```suggestion
   * the appearance of a different AWS Region in the encryption context can indicate malicious interference.
```

File: modules/example-browser/src/caching_materials_manager.ts [hunk 20]
```diff
@@ -0,0 +1,201 @@
...
+   * See: https://docs.aws.amazon.com/encryption-sdk/latest/developer-guide/concepts.html#encryption-context
+   *
+   * DATA KEYS FOR MESSAGES WILL ***ONLY*** BE SHARED FOR EXACT MATCHES OF ENCRYPTION CONTEXT.
```
Comment (juneb) [inline modules/example-browser/src/caching_materials_manager.ts]:
```suggestion
   * Cached data keys are reused ***only*** when the encryption contexts passed into the functions are an exact case-sensitive match.  
```

File: modules/example-browser/src/caching_materials_manager.ts [hunk 21]
```diff
@@ -0,0 +1,201 @@
...
+  const plainText = new Uint8Array([1, 2, 3, 4, 5])
+
+  /* Encrypt the data.
+   * The caching CMM only reuses data keys
+   * when it know the length (or an estimate) of the plaintext.
+   * However, in the browser,
+   * you must provide all of the plaintext to the encrypt function.
+   * Therefore, the encrypt function in the browser knows the length of the plaintext
+   * and does not accept a plaintextLength option.
+   */
```
Comment (juneb) [inline modules/example-browser/src/caching_materials_manager.ts]: Ryan, is there a reason to explain why you can't provide a value? This function doesn't have this parameter. I would just omit lines 148-153. (Feel free to overrule me if I've misinterpreted this.)

File: modules/example-browser/src/multi_keyring.ts [hunk 22]
```diff
@@ -62,7 +63,7 @@ declare const AWS_CREDENTIALS: {accessKeyId: string, secretAccessKey:string }
    * You should inject your credential into the browser in a secure manner
```
Comment (juneb) [inline modules/example-browser/src/multi_keyring.ts]:
```suggestion
   * You should inject your credentials into the browser in a secure manner
```

File: modules/example-browser/src/caching_materials_manager.ts (hunk 21 above)
Comment (seebees) [reply modules/example-browser/src/caching_materials_manager.ts]: If you provide a value and it does not match, this just makes for confusion.

File: modules/example-browser/src/caching_cmm.ts [hunk 23]
```diff
@@ -0,0 +1,215 @@
...
+   * so that you can try decrypting it with another AWS Encryption SDK implementation.
+   */
+  const resultBase64 = toBase64(result)
+  console.log(resultBase64)
+  document.write(resultBase64)
+
+  /* Decrypt the data.
+   * NOTE: This decrypt request will not use the data key
+   * that was cached during the encrypt operation.
+   * Data keys for encrypt and decrypt operations are cached separately.
```
Comment (juneb) [inline modules/example-browser/src/caching_cmm.ts]: Just a question. Are the data keys cached separately or just not reused when the operations differ. 

Should I add "operation (encrypt, decrypt)" to this section that explains when cached data keys are reused?
https://docs.aws.amazon.com/encryption-sdk/latest/developer-guide/data-caching-details.html#how-caching-works

File: modules/example-browser/src/caching_cmm.ts [hunk 24]
```diff
@@ -0,0 +1,215 @@
...
+  const { plaintext, messageHeader } = await decrypt(cachingCMM, result)
+
+  /* Grab the encryption context so you can verify it. */
+  const { encryptionContext: decryptedContext } = messageHeader
+
+  /* Verify the encryption context.
+   * If you use an algorithm suite with signing,
+   * the Encryption SDK adds a name-value pair to the encryption context that contains the public key.
+   * Because the encryption context might contain additional key-value pairs,
+   * do not add a test that requires that all key-value pairs match.
```
Comment (juneb) [inline modules/example-browser/src/caching_cmm.ts]:
```suggestion
   * do not include a test that requires that all key-value pairs match.
```

File: modules/example-browser/src/caching_cmm.ts [hunk 25]
```diff
@@ -0,0 +1,215 @@
...
+   * Instead, verify that the key-value pairs you expect match.
```
Comment (juneb) [inline modules/example-browser/src/caching_cmm.ts]:
```suggestion
   * Instead, verify that the key-value pairs that you supplied to the `encrypt` function are included in the encryption context that the `decrypt` function returns.
```

File: modules/example-node/src/caching_cmm.ts [hunk 26]
```diff
@@ -0,0 +1,147 @@
...
+  /* A KMS CMK is required to generate the data key.
+   * You need kms:GenerateDataKey permission on the CMK in generatorKeyId.
+   */
+  const generatorKeyId = 'arn:aws:kms:us-west-2:658956600833:alias/EncryptDecrypt'
+
+  /* Adding alternate KMS keys that can decrypt.
+   * Access to kms:Encrypt is required for every CMK in keyIds.
+   * You might list several keys in different AWS Regions.
+   * This allows you to decrypt the data in any of the represented Regions.
+   * In this example, I am using the same CMK.
```
Comment (juneb) [inline modules/example-node/src/caching_cmm.ts]: This should match the text in the browser example.

File: modules/example-node/src/caching_cmm.ts [hunk 27]
```diff
@@ -0,0 +1,147 @@
...
+   * This is *only* to demonstrate how the CMK ARNs are configured.
+   */
+  const keyIds = ['arn:aws:kms:us-west-2:658956600833:key/b3537ef1-d8dc-4780-9f5a-55776cbb2f7f']
+
+  /* The KMS keyring must be configured with the desired CMKs
+   * In this case we are going to pass this keyring to the caching CMM,
```
Comment (juneb) [inline modules/example-node/src/caching_cmm.ts]:
```suggestion
   * This example passes the keyring to the caching CMM
```

File: modules/example-node/src/caching_cmm.ts [hunk 28]
```diff
@@ -0,0 +1,147 @@
...
+   * instead of using it directly.
+   */
+  const keyring = new KmsKeyringNode({ generatorKeyId, keyIds })
+
+  /* Create a cache to hold the material.
+   * In this case we use the local cache provided by the Encryption SDK.
```
Comment (juneb) [inline modules/example-node/src/caching_cmm.ts]: Please make all text in the node example match the corresponding edited text in the browser example.

File: modules/example-browser/src/aes_simple.ts [hunk 29]
```diff
@@ -27,10 +27,11 @@ import {
...
 
-  ;(async function testAES () {
+/* This is done to facilitate testing. */
+export async function testAES () {
   /* You need to specify a name
    * and a namespace for raw encryption key providers.
-   * The name and namespace that you use in the decryption keyring *must* be an exact,
+   * The name and namespace that you use in the decryption keyring *must* be an exact, case-sensitive match.
    * *case-sensitive* match for the name and namespace in the encryption keyring.
    */
```
[the remark this answers is not in the document]
Comment (seebees) [reply modules/example-browser/src/aes_simple.ts]: fixed

File: modules/example-browser/src/caching_cmm.ts (hunk 23 above)
Comment (seebees) [reply modules/example-browser/src/caching_cmm.ts]: The are cached under separate ID's.  So they are in the same cache, but they are intentionally not shared.

File: modules/client-node/Readme.md [hunk 30]
```diff
@@ -61,10 +61,10 @@ const cleartext = 'asdf'
  * the Encryption SDK returns an "encrypted message" that includes the ciphertext,
```
[the remark this answers is not in the document]
Comment (seebees) [reply modules/client-node/Readme.md]: done

File: modules/example-node/src/caching_cmm.ts (hunk 26 above)
Comment (seebees) [reply modules/example-node/src/caching_cmm.ts]: fixed

File: modules/example-node/src/caching_cmm.ts (hunk 28 above)
Comment (seebees) [reply modules/example-node/src/caching_cmm.ts]: fixed

**Pull request merged**